smolyay/surrogate.py

```diff
@@ -0,0 +1,279 @@
+        self._coefficients = None
+
+    @property
+    def domain(self):
+        """list: domain of the real function to be approximated."""
+        return self._domain
+
+    @domain.setter
+    def domain(self, domain_real_function):
+        self._domain = domain_real_function
+
+    @property
+    def grids_generator(self):
+        """:class:`IndexGridGenerator`: Generator used for making the grids."""
+        return self._grids_generator
+
+    @grids_generator.setter
+    def grids_generator(self, grids_generator):
+        if not isinstance(grids_generator, IndexGridGenerator):
+            raise TypeError('Grids must be IndexGridGenerator')
+        self._grids_generator = grids_generator
+
+    def _make_basis_matrix(self):
+        """Generate basis matrix."""
+        self._basis_matrix = numpy.ones((len(self._grids.points),
+                                         len(self._grids.points)))
+        point_num = 0
+        for points in self._grids.points:
+            for dimension_ in range(self._dimension):
+                grid_num = 0
+                for basis_function in list(zip(
+                        *self._grids.basis_functions))[dimension_]:
+                    self._basis_matrix[point_num][grid_num] *= (
+                        basis_function(points[dimension_]))
+                    grid_num += 1
+            point_num += 1
+
+    def train(self, real_function, linear_solver='lu'):
+        """Fit surrogate's components (basis functions) to the function.
+
+        Parameters
+        ----------
+        real_function: callabe
+            Function to be approximated.
+        linear_slover: string
+            Method of solving linear equations:
+                'lu': lower upper decomposition
+                'inv': solve for coefficients through x = A^{-1}B.
+                'lstsq': least-square solution.
+
+        Returns
+        -------
+        list
+            Coefficients of the surrogate.
+        """
+        self.real_function = real_function
+        self.linear_solver = linear_solver
+        self._make_basis_matrix()
+
+        # transform grid points
+        transformed_grids = numpy.zeros((
+            len(self._grids.points), self._dimension))
+        for dimension_ in range(self._dimension):
+            transformed_grids[:, dimension_] = (
+                numpy.polynomial.polyutils.mapdomain(
+                    numpy.array(self._grids.points)[:, dimension_],
+                    (-1, 1), self.domain[dimension_]))
+
+        # evaluate real function at grid points
+        real_function_at_grids = numpy.zeros((len(self._grids.points)))
+        for point_num in range(len(self._grids.points)):
+            real_function_at_grids[point_num] = (
+                self.real_function(*transformed_grids[point_num]))
+
+        # fit
+        self._coefficients = solve_linear(
+            self._basis_matrix, real_function_at_grids,
+            self.linear_solver)
```

Comment:
I don't quite understand why this can't be:
```suggestion
        # fit
        self.train_from_data(real_function_at_grids, linear_solver)
```